### PR TITLE
JDK-8269092: Add OldObjectSampleEvent.allocationSize field

### PR DESCRIPTION
**Pull request by luhenry**
The data is already available in ObjectSample.allocated and getting this data
in another way is complicated (via Instrumentation for example). Having the
size of the objects also allows to augment the memory leak analysis based on
the object size impact on top of based on the object count impact.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269092](https://bugs.openjdk.java.net/browse/JDK-8269092): Add OldObjectSampleEvent.allocationSize field

### Reviewers
 * [Erik Gahlin](https://openjdk.java.net/census#egahlin) (@egahlin - **Reviewer**)
 * [Jaroslav Bachorik](https://openjdk.java.net/census#jbachorik) (@jbachorik - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4549/head:pull/4549` \
`$ git checkout pull/4549`

Update a local copy of the PR: \
`$ git checkout pull/4549` \
`$ git pull https://git.openjdk.java.net/jdk pull/4549/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4549`

View PR using the GUI difftool: \
`$ git pr show -t 4549`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4549.diff">https://git.openjdk.java.net/jdk/pull/4549.diff</a>

</details>

### CODE REVIEW

**Comment by egahlin** [comment]
How will the "allocation size" help in the analysis?

I think this likely to confuse people more than actually help people find memory leaks.

**Comment by luhenry** [comment]
Right now the only information about the OldObject in this sample we have are the age, time of the allocation, the number of elements in the array (if it's one), the total amount of heap used (overall, for the whole process, not just this object), and the class, address, and referrer of the object itself.

In the case of memory leak analysis, we then know which objects appear to be leaking. We also know that, based on the heuristics of the priority queue in the leak profiler, the sampled objects have the most impact on memory usage (since sampling is biased towards bigger shallow objects). However, we still don't know exactly how much memory is used. We can approximate it if it's an array based on the size of the elements and the number of elements. However, it's always going to be approximative since the size of the object header might change in the future (project Lilliput), and because of various other constraints like alignement of elements, size of each element, etc.

I'm sure these constraints would not necessarily be impossible to overcome, but because we already have this exact information, I figured it to be much easier to put it straight in the event. This exact field is also present in `AllocationRequiringGC`, `ObjectAllocationInNewTLAB`, and `ObjectAllocationOutsideTLAB`, so there is precedent in terms of naming, typing, and meaning for related events.

**Comment by egahlin** [comment]
> In the case of memory leak analysis, we then know which objects appear to be leaking. We also know that, based on the heuristics of the priority queue in the leak profiler, the sampled objects have the most impact on memory usage (since sampling is biased towards bigger shallow objects).

Not sure I understand, but it may not matter

> However, we still don't know exactly how much memory is used. We can approximate it if it's an array, based on the size of the elements and the number of elements. However, it's always going to be approximative since the size of the object header might change in the future (project Lilliput), and because of various other constraints like alignement of elements, size of each element, etc.

But why do you need it? 

Is a large (or small?) "allocation size" more likely a leak, or what is the thinking?

**Comment by luhenry** [comment]
> Is a large (or small?) "allocation size" more likely a leak, or what is the thinking?

It's not so much that it's more likely to leak, but it's to allow the user to prioritise accordingly to the size of the leak. The bigger the leak, the more urgent it seems because the more impactful it is to fix. Overall, it doesn't hurt to have this information, it's trivial to gather (it's a one-liner), and it can help users of JFR to prioritise which leak to fix.

**Comment by luhenry** [comment]
Let me just add a quick example of how we plan to use it.

At Datadog, we have for other languages Heap Live Objects and Heap Live Size profilers. These work for Python and Go in a very similar manner to the OldObjectSample in JFR: it takes a sample of objects, and track their lifetime.

Here is an example of what it looks like in practice on a sample profile:

image
image

JFR generally tracks more useful data, like the reference chains. Unfortunately, we are still blocked for the Heap Live Size profiler in Java because we need to match an OldObjectSample to the size it occupies on the heap, and we have no way today to map precisely the object type to the object allocated size. This is what this PR is attempting to fix.

**Comment by egahlin** [comment]
> Let me just add a quick example of how we plan to use it.

Ok, thanks for the explanation.

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 1]
```diff
@@ -627,6 +627,7 @@
   <Event name="OldObjectSample" category="Java Virtual Machine, Profiling" label="Old Object Sample" description="A potential memory leak" stackTrace="true" thread="true"
     startTime="false" cutoff="true">
     <Field type="Ticks" name="allocationTime" label="Allocation Time" />
+    <Field type="ulong" contentType="bytes" name="allocationSize" label="Allocation Size" />
```
Comment (egahlin) [inline src/hotspot/share/jfr/metadata/metadata.xml:630]: I wonder if the label should be "Object Size"?

File: test/jdk/jdk/jfr/event/oldobject/TestAllocationSize.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                leak.add(new Leak2());
+                leak.add(new Leak3());
+            }
+
+            recording.stop();
+
+            List<RecordedEvent> events = Events.fromRecording(recording);
+            Events.hasEvents(events);
+            for (RecordedEvent e : events) {
+                if (e.getEventType().getName().equals(EventNames.OldObjectSample)) {
```
Comment (egahlin) [inline test/jdk/jdk/jfr/event/oldobject/TestAllocationSize.java:81]: No need to check against event name. 

If only OldObjectSample event has been enable, other events won't be there.

File: test/jdk/jdk/jfr/event/oldobject/TestAllocationSize.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                    RecordedObject object = e.getValue("object");
+                    RecordedClass type = object.getValue("type");
+                    if ((long)e.getValue("allocationSize") <= 0) {
```
Comment (egahlin) [inline test/jdk/jdk/jfr/event/oldobject/TestAllocationSize.java:84]: You could use e.getLong(...) here.

**Comment by luhenry** [comment]
I've updated according to your review and slightly enhanced the test case.

**Review by egahlin — APPROVED** [review]

**Review by jbachorik — APPROVED** [review]

**Pull request closed without merging**